modules/hal_infineon/mtb-pdl-cat1/CMakeLists.txt

```diff
@@ -0,0 +1,55 @@
+# Copyright (c) 2022 Cypress Semiconductor Corporation.
+#
+# SPDX-License-Identifier: Apache-2.0
+
+set(pdl_dir             ${ZEPHYR_HAL_INFINEON_MODULE_DIR}/mtb-pdl-cat1)
+set(pdl_drv_dir         ${ZEPHYR_HAL_INFINEON_MODULE_DIR}/mtb-pdl-cat1/drivers)
+set(pdl_dev_cat1a_dir   ${ZEPHYR_HAL_INFINEON_MODULE_DIR}/mtb-pdl-cat1/devices/COMPONENT_CAT1A)
+
+zephyr_compile_definitions(${CONFIG_SOC_PART_NUMBER})
+
+# Add mtb-pdl-cat1
+zephyr_include_directories(${pdl_drv_dir}/include)
+zephyr_include_directories(${pdl_dev_cat1a_dir}/include)
+zephyr_include_directories(${pdl_dev_cat1a_dir}/include/ip)
+zephyr_include_directories(${pdl_dev_cat1a_dir}/templates/COMPONENT_MTB)
+
+zephyr_library_sources(${pdl_dev_cat1a_dir}/source/cy_device.c)
+zephyr_library_sources(${pdl_drv_dir}/source/TOOLCHAIN_GCC_ARM/cy_syslib_gcc.S)
```

Comment:
those functions are used by PDL functions.

similar in another vendors: 
__get_PRIMASK/__set_PRIMASK - in stm32
DisableGlobalIRQ/EnableGlobalIRQ - in nxp

Reply:
Yes, this is literally `irq_lock()` and `irq_unlock()` in Zephyr

Reply:
I see many drivers making use of `Cy_SysLib_EnterCriticalSection()` and `Cy_SysLib_ExitCriticalSection()`, could you please explain why? could that have an impact when using ZLI?

Reply:
@npal-cy @gmarull Right now I’d go with the Infineon critical sections until you can figure out how to abstract them, like nrfx does:
https://github.com/zephyrproject-rtos/zephyr/blob/main/modules/hal_nordic/nrfx/nrfx_glue.h

Reply:
The drivers use critical section during operation with hw, where the sequences is critical and can not be interrupted by side interrupt event. 
Regarding to impact, it really depends on use case.  The driver's run-time functions can uses critical section to cover some register access (short time operation), some complex sequences under critical section are mostly in init functions, which should not impact ZLI.

Reply:
To be honest, I'm not happy on introducing HALs that do things _their own way_. There should probably be some rules on what HALs can/cannot do. Otherwise, things like https://github.com/zephyrproject-rtos/zephyr/pull/42395 happen, ie working around HALs in Zephyr. Similar things happen for clock control: many SoCs call a "init" function from the HAL that does all the magic, bypassing any kind of good practices.

To summarize, low level HALs _can_ be useful, higher level ones tend to be inefficient and cause troubles because they do _too much_. This is likely because they were designed in a bare-metal/FreeRTOS mindset, where you basically have no infrastructure.

Modules are nowadays an entry point for code that doesn't get the same level of review that upstream code does, but still, often it is code that is fundamental to build an image.

Reply:
@gmarull, i do not agree about you concerns for vendor implementation of critical section. We are talking about PDL - low level layer, which directly works with HW, and knows when need to use critical section for particular specific HW and how implement in to achieve all needs.. 

Zephyr can not require to use Zephyr implementation (irq_lock/irq_unlock) for low level (vendor specific) layer, because it does not know all HW specifics.. Zephyr can require to do this for Zephyr drivers, app, or other modules, which are upper vendor peripheral drivers (low level).

I can not guarantee that our HW will work properly with irq_lock,  because irq_lock allows to interrupt the locked by irq_lock  code if ZLI enabled (it uses BASEPRI instead PRIMASK).

Reply:
So what will happen with the current implementation of the `Cy_SysLib_*CriticalSection` functions, which just use the Armv6-M `cpsid/e` basic `PRIMASK` access is that:

- Infineon HAL drivers will disable/enable ZLIs. This is unfortunate, but you could just add a Kconfig that disallows `CONFIG_ZERO_LATENCY_IRQS` with this SoC
- Infineon HAL drivers will disable the highest prio (prio 0) which is used by Zephyr's fault handling code.  This means that if a fault is triggered inside one of those critical sections then it will not be handled

All in all I don't see this as a show stopper to get this merged, but it certainly needs to be fixed at some point in the near future.